ocaml/testsuite/tests/typing-layouts-bits32/c_functions.c

```diff
@@ -0,0 +1,50 @@
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/alloc.h>
+#include <caml/custom.h>
+
+CAMLprim int32_t lognot_UtoU(int32_t u) {
+  return ~u;
+}
+
+CAMLprim int32_t lognot_BtoU(value u) {
+  return ~Int32_val(u);
```

Comment:
I love the idea that I am the expert you call for to answer questions about the FFI.

I think the second thing you said is right.  The manual has other examples later that do this (e.g., `foo_byte`).  And in one case it observes "Here, the registering of result is not strictly needed, because no allocation takes place after it gets its value, but it’s easier and safer to simply register all the local variables that have type value."  But I agree it would be nicer if the actual rules were written down somewhere.

Reply:
This code is fine!